### PR TITLE
Add getBlockConfidence rpc endpoint

### PR DESCRIPTION
**Pull request by CriesofCarrots**
#### Problem
The RPC API returns results based on either the latest bank, or a bank from a couple blocks ago. But what clients really want is results based on how much of the cluster has voted on the results.

Each validator maintains a confidence cache -- ie. a table of how much stake has voted for a particular fork at slot depths back through root -- but there is no way for clients to access this data.

#### Summary of Changes
- Passes ForkConfidenceCache into JsonRpcRequestProcessor
- Adds getBlockConfidence rpc endpoint to expose raw confidence data to clients

Toward #6346 
This is foundational work. Next steps are to build mechanisms to examine the cache to find the latest block that meets confidence criteria (default: rooted, 2/3 of stake voted; customizable by client) and use that bank-fork for RPC results.

### CODE REVIEW

**Comment by mvines** [comment]
So https://docs.solana.com/book/terminology#fork makes this pretty confusing to me.  I think what we're really talking about is block confidence right?  `getBlockConfidence`.  

Also this:
image

File: book/src/api-reference/jsonrpc-api.md [hunk 1]
```diff
@@ -202,6 +203,34 @@ curl -X POST -H "Content-Type: application/json" -d '{"jsonrpc":"2.0","id":1, "m
...
+* 'integer' - total stake in the cluster
+
+#### Example:
+
+```bash
+// Request
+curl -X POST -H "Content-Type: application/json" -d '{"jsonrpc":"2.0","id":1, "method":"getForkConfidence","params":[5]}' http://localhost:8899
+
+// Result
+{"jsonrpc":"2.0","result":{"epoch":3,"slotIndex":126,"slotsInEpoch":256},"id":1}
```
Comment (mvines) [inline book/src/api-reference/jsonrpc-api.md]: this result output looks wrong?

Comment (CriesofCarrots) [reply book/src/api-reference/jsonrpc-api.md]: Sure is!

**Comment by CriesofCarrots** [comment]
> So https://docs.solana.com/book/terminology#fork makes this pretty confusing to me. I think what we're really talking about is block confidence right? `getBlockConfidence`.

Yeah, that's fair. I'll reword.

File: book/src/api-reference/jsonrpc-api.md [hunk 2]
```diff
@@ -149,6 +150,34 @@ curl -X POST -H "Content-Type: application/json" -d '{"jsonrpc":"2.0", "id":1, "
 {"jsonrpc":"2.0","result":0,"id":1}
 ```
 
+### getBlockConfidence
+
+Returns confidence for particular block
+
+#### Parameters:
+
+* `u64` - fork, identified by slot id
```
Comment (mvines) [inline book/src/api-reference/jsonrpc-api.md]: `block, identified by block id`?

File: book/src/api-reference/jsonrpc-api.md [hunk 3]
```diff
@@ -149,6 +150,34 @@ curl -X POST -H "Content-Type: application/json" -d '{"jsonrpc":"2.0", "id":1, "
...
+
+#### Results:
+
+The result field will be an array with two fields:
+
+* Confidence
+  * `null` - Unknown fork
```
Comment (mvines) [inline book/src/api-reference/jsonrpc-api.md]: `block`

File: book/src/api-reference/jsonrpc-api.md [hunk 4]
```diff
@@ -149,6 +150,34 @@ curl -X POST -H "Content-Type: application/json" -d '{"jsonrpc":"2.0", "id":1, "
...
+  * `object` - BankConfidence
+    * `array` - confidence, array of u64 integers logging the amount of stake in the cluster that has voted on the block at a particular depth from 0 to `MAX_LOCKOUT_HISTORY`
+* 'integer' - total stake in the cluster
```
Comment (mvines) [inline book/src/api-reference/jsonrpc-api.md]: total stake *in lamports, at that block*?

File: book/src/api-reference/jsonrpc-api.md (hunk 2 above)
Comment (CriesofCarrots) [reply book/src/api-reference/jsonrpc-api.md]: That's not quite right. Block id is the last entry id (hash) in the block, and this is just returning the Slot.

File: book/src/api-reference/jsonrpc-api.md (hunk 4 above)
Comment (CriesofCarrots) [reply book/src/api-reference/jsonrpc-api.md]: I don't think that's quite right. It's the total stake the validator currently sees in the cluster, not the stake that was delegated at the time of the block. Can you thin of a better way to put that?

**Comment by CriesofCarrots** [comment]
> I'm trying to think how I'd use `getBlockConfidence`. I think I'd generally have a transaction signature and I just want to know how much of the cluster has voted on the fork that contains it. "Is my transaction 2/3+ confirmed or not?"

You wouldn't use this endpoint for that. That use case is what my next PRs are going to implement.
I just thought it would be useful to have a way to see the raw confidence data for every slot.

File: book/src/api-reference/jsonrpc-api.md (hunk 2 above)
Comment (mvines) [reply book/src/api-reference/jsonrpc-api.md]: ah, k.  So if we have a virtual slot then we'd end up with a `null` response?

File: book/src/api-reference/jsonrpc-api.md (hunk 4 above)
Comment (mvines) [reply book/src/api-reference/jsonrpc-api.md]: how about: `total active stake, in lamports, of the current epoch`

File: book/src/api-reference/jsonrpc-api.md (hunk 2 above)
Comment (CriesofCarrots) [reply book/src/api-reference/jsonrpc-api.md]: What's a virtual slot?

**Comment by mvines** [comment]
> You wouldn't use this endpoint for that. That use case is what my next PRs are going to implement.
> I just thought it would be useful to have a way to see the raw confidence data for every slot on command.

Got it!  Apparently I didn't read your entire PR description 😊

File: book/src/api-reference/jsonrpc-api.md (hunk 2 above)
Comment (mvines) [reply book/src/api-reference/jsonrpc-api.md]: A slot that produced no block.

File: book/src/api-reference/jsonrpc-api.md (hunk 4 above)
Comment (CriesofCarrots) [reply book/src/api-reference/jsonrpc-api.md]: Definitely better!

**Comment by mvines** [comment]
> ...Done. I was reluctant to go deeper with the renaming (into confidence.rs), because replay_stage uses "fork" this way a lot vis-a-vis BankForks. Does it seem reasonable, @mvines ?

Yep.  Although I'd love to clean up the internal naming eventually too.  We use `fork` and `fork_id` interchangeably too internally :-/

File: book/src/api-reference/jsonrpc-api.md (hunk 2 above)
Comment (CriesofCarrots) [reply book/src/api-reference/jsonrpc-api.md]: Gotcha! Yes, that should be correct, because such a slot would not be cached in the ForkConfidenceCache

**Pull request merged**